File: go/core/tool_hook.go
package core

import "context"

// ToolResult is the outcome of a single tool call. It is handed to
// ToolHook.PostCall as a mutable pointer so a hook can rewrite Content before it
// reaches the model — the redaction seam (e.g. scrubbing a leaked secret from a
// tool's output). Mirrors the Rust engine's ToolResult.
type ToolResult struct {
	// ToolCallID is the id of the ToolCall this result answers.
	ToolCallID string
	// Content is the text the model sees. A PostCall hook may mutate it.
	Content string
	// IsError is true when the tool failed (or a PreCall hook blocked it).
	IsError bool
}

// ToolHook observes and can gate/redact tool calls, mirroring the Rust engine's
// ToolHook lifecycle (pre_call / post_call). Hooks are registered via
// AgentOptions.Hooks and run in registration order around every dispatched tool
// call. A hook must be safe for concurrent use: with ParallelToolCalls the
// engine may run PreCall/PostCall from multiple goroutines at once.
type ToolHook interface {
	// PreCall runs before a tool executes. Returning a non-nil error blocks the
	// call: the tool never runs and the model is told it was blocked.
	PreCall(ctx context.Context, call ToolCall) error
	// PostCall runs after a tool executes, with a mutable *ToolResult it may
	// rewrite (the redaction seam). A returned error is ignored — the
	// (possibly redacted) result still reaches the model — matching the Rust
	// engine, where a post-hook failure is logged, not surfaced.
	PostCall(ctx context.Context, call ToolCall, result *ToolResult) error
}

File: dotnet/core/src/ToolHook.cs
using Microsoft.Extensions.AI;

namespace SmooAI.SmoothOperator.Core;

/// <summary>
/// A hook that runs around every tool call — the in-process surveillance / redaction seam the
/// smooth-operator engines share. Mirrors the Rust reference's <c>ToolHook</c> trait
/// (<c>rust/smooth-operator-core/src/tool.rs</c>): <see cref="PreCallAsync"/> before the tool runs,
/// <see cref="PostCallAsync"/> after — with a <b>mutable</b> result so a hook can redact.
///
/// The engine speaks Microsoft.Extensions.AI types throughout, so a hook does too: the call is the
/// model's <see cref="FunctionCallContent"/> and the result is the <see cref="FunctionResultContent"/>
/// fed back to the model. Both methods default to a no-op, so a hook overrides only the phase it
/// cares about (e.g. a redactor implements only <see cref="PostCallAsync"/>).
/// </summary>
public interface IToolHook
{
    /// <summary>
    /// Called before the tool executes. <b>Throw to block the call</b> — parity with the Rust
    /// reference's <c>pre_call</c> returning <c>Err</c>. When a hook throws, the tool never runs
    /// and the model is fed a "Blocked by hook" result instead. The default impl is a no-op.
    /// </summary>
    Task PreCallAsync(FunctionCallContent call, CancellationToken cancellationToken = default) => Task.CompletedTask;

    /// <summary>
    /// Called after the tool executes, with the <b>mutable</b> <see cref="FunctionResultContent"/>.
    /// This is a redaction seam, not just an observation point: a hook may rewrite
    /// <see cref="FunctionResultContent.Result"/> (e.g. to scrub a leaked secret) and the mutation is
    /// what the caller — and the model/conversation — sees. Runs on both successful and errored tool
    /// results. A throw here is swallowed (the possibly-redacted result still reaches the caller),
    /// parity with the Rust reference logging <c>post_call</c> errors rather than surfacing them. The
    /// default impl is a no-op.
    /// </summary>
    Task PostCallAsync(FunctionCallContent call, FunctionResultContent result, CancellationToken cancellationToken = default) => Task.CompletedTask;
}
